runtime/expr/function/flatten.go

```diff
@@ -40,12 +44,35 @@ func (n *Flatten) Call(ctx zed.Allocator, args []zed.Value) *zed.Value {
 		}
 		n.mapper.EnterType(typ.ID(), valType)
 	}
-	if n.keyType == nil {
-		n.keyType = n.zctx.LookupTypeArray(zed.TypeString)
-	}
 	n.Reset()
 	n.encode(typ.Columns, valType, field.Path{}, val.Bytes)
-	return ctx.NewValue(n.zctx.LookupTypeMap(n.keyType, valType), n.Bytes())
+	return ctx.NewValue(n.zctx.LookupTypeArray(valType), n.Bytes())
+}
+
+func (n *Flatten) collectTypes(cols []zed.Column) []zed.Type {
+	var types []zed.Type
+	for _, col := range cols {
+		if typ := zed.TypeRecordOf(col.Type); typ != nil {
+			for _, typ := range n.collectTypes(typ.Columns) {
+				types = append(types, typ)
+			}
+		} else {
+			typ, ok := n.entryTypes[col.Type.ID()]
+			if !ok {
+				var err error
+				typ, err = n.zctx.LookupTypeRecord([]zed.Column{
+					zed.NewColumn("key", n.keyType),
+					zed.NewColumn("value", col.Type),
+				})
+				if err != nil {
+					panic(fmt.Errorf("flatten: error creating record type %w", err))
+				}
+				n.entryTypes[col.Type.ID()] = typ
```

Comment:
```suggestion
				n.entryTypes[col.Type] = n.zctx.MustLookupTypeRecord([]zed.Column{
					zed.NewColumn("key", n.keyType),
					zed.NewColumn("value", col.Type),
				})
```

Reply:
This doesn't set `typ` but I'll fix this.